### PR TITLE
Upcycling support for GPT based model to a MoE model

### PR DESCRIPTION
**Pull request by adityavavre**
# What does this PR do ?

Adds support for upcycling pretrained weights to a MoE model. This follows the paper: SPARSE UPCYCLING: TRAINING MIXTURE-OF-EXPERTS FROM DENSE CHECKPOINTS ([https://arxiv.org/pdf/2212.05055](https://arxiv.org/pdf/2212.05055))

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
```python
python NeMo/examples/nlp/language_modeling/megatron_gpt_upcycling.py \
--config-path=/path/to/MoE/config/ \
--config-name=MoE/config/name.yaml \
+restore_from_path=/path/to/base/model 
```
Where you MoE config should contain the target MoE model you want to train. The base model you want to upcycle can also be directly added to this config under the argument `restore_from_path`. 

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by ethanhe42** [comment]
could you also provide the instruction to run the script?

**Review by ethanhe42 — APPROVED** [review]
